boot/freeldr/freeldr/lib/fs/iso.c

```diff
@@ -140,116 +171,151 @@ static ARC_STATUS IsoBufferDirectory(ULONG DeviceId, ULONG DirectoryStartSector,
+    else
+    {
+        PathPart = PathBuffer;
+        DoFullLookup = TRUE;
+
+        DirectorySector = Volume->PvdDirectorySector;
+        DirectoryLength = Volume->PvdDirectoryLength;
+
+        /* Skip leading path separator, if any */
+        if (*FileName == '\\' || *FileName == '/')
+            ++FileName;
+
+        /* Figure out how many sub-directories we are nested in */
+        NumberOfPathParts = FsGetNumPathParts(FileName);
+    }
 
     //
     // Loop once for each part
     //
     for (i=0; i<NumberOfPathParts; i++)
     {
-        //
-        // Get first path part
-        //
-        FsGetFirstNameFromPath(PathPart, FileName);
-
-        //
-        // Advance to the next part of the path
-        //
-        for (; (*FileName != '\\') && (*FileName != '/') && (*FileName != '\0'); FileName++)
+        if (DoFullLookup)
         {
-        }
-        FileName++;
+            /* Get first path part */
+            FsGetFirstNameFromPath(PathPart, FileName);
 
-        //
-        // Buffer the directory contents
-        //
-        Status = IsoBufferDirectory(DeviceId, DirectorySector, DirectoryLength, &DirectoryBuffer);
-        if (Status != ESUCCESS)
-            return Status;
+            /* Advance to the next part of the path */
+            for (; (*FileName != '\\') && (*FileName != '/') && (*FileName != '\0'); FileName++)
+            {
+            }
+            FileName++;
+
+            /* Buffer the directory contents */
+            Status = IsoBufferDirectory(DeviceId,
+                                        DirectorySector,
+                                        DirectoryLength,
+                                        &DirectoryBuffer);
+            if (Status != ESUCCESS)
+                return Status;
+
+            /* Save the directory buffer */
+            if ((i + 1) >= NumberOfPathParts)
+            {
+                if (Volume->DirectoryBuffer)
+                {
+                    FrLdrTempFree(Volume->DirectoryBuffer, TAG_ISO_BUFFER);
+                    Volume->DirectoryBuffer = NULL;
+                }
 
-        //
-        // Search for file name in directory
-        //
-        if (!IsoSearchDirectoryBufferForFile(DirectoryBuffer, DirectoryLength, PathPart, &IsoFileInfo))
+                Volume->DirectoryPathLength = DirectoryPathLength;
+                Volume->DirectoryBuffer = DirectoryBuffer;
+                Volume->DirectoryLength = DirectoryLength;
+
+                RtlCopyMemory(Volume->DirectoryPath, FullPath, DirectoryPathLength);
+            }
+        }
+
+        /* Search for file name in directory */
+        if (!IsoSearchDirectoryBufferForFile(DirectoryBuffer,
+                                             DirectoryLength,
+                                             PathPart,
+                                             IsoFileInfo))
         {
-            FrLdrTempFree(DirectoryBuffer, TAG_ISO_BUFFER);
+            if ((i + 1) < NumberOfPathParts)
```

Comment:
The DirectoryBuffer should only be freed conditionally, or not?

Reply:
The cached DirectoryBuffer can't be freed.